cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -359,34 +365,69 @@ public void stop(ProgramId programId, @Nullable String runId) throws Exception {
    * @throws UnauthorizedException if the user issuing the command is not authorized to stop the program. To stop a
    *                               program, a user requires {@link Action#EXECUTE} permission on the program.
    */
-  public List<ListenableFuture<ProgramController>> issueStop(ProgramId programId, @Nullable String runId)
-    throws Exception {
+  public List<ListenableFuture<ProgramController>> issueStop(ProgramId programId,
+                                                             @Nullable String runId) throws Exception {
     authorizationEnforcer.enforce(programId, authenticationContext.getPrincipal(), Action.EXECUTE);
-    List<ProgramRuntimeService.RuntimeInfo> runtimeInfos = findRuntimeInfo(programId, runId);
-    if (runtimeInfos.isEmpty()) {
+
+    // See if the program is running as per the runtime service
+    Map<RunId, RuntimeInfo> runtimeInfos = findRuntimeInfo(programId, runId);
+    Map<ProgramRunId, RunRecordMeta> activeRunRecords = getActiveRuns(programId, runId);
+
+    if (runtimeInfos.isEmpty() && activeRunRecords.isEmpty()) {
+      // Error out if no run information from runtime service and from run record
       if (!store.applicationExists(programId.getParent())) {
         throw new ApplicationNotFoundException(programId.getParent());
       } else if (!store.programExists(programId)) {
         throw new ProgramNotFoundException(programId);
-      } else if (runId != null) {
-        ProgramRunId programRunId = programId.run(runId);
-        // Check if the program is running and is started by the Workflow
-        RunRecordMeta runRecord = store.getRun(programId, runId);
+      }
+      throw new BadRequestException(String.format("Program '%s' is not running.", programId));
+    }
+
+    // Stop the running program based on a combination of runtime info and run record
+    // It's possible that some of them are not yet available from the runtimeService due to timing
+    // differences between the run record was created vs being added to runtimeService
+    // So we retry in a loop for up to 3 seconds max to cater for those cases
+
+    Set<String> pendingStops = Stream.concat(runtimeInfos.keySet().stream().map(RunId::getId),
+                                             activeRunRecords.keySet().stream().map(ProgramRunId::getRun))
+                                      .collect(Collectors.toSet());
+
+    List<ListenableFuture<ProgramController>> futures = new ArrayList<>();
+    Stopwatch stopwatch = new Stopwatch().start();
+
+    while (!pendingStops.isEmpty() && stopwatch.elapsedTime(TimeUnit.SECONDS) < 3L) {
+      Iterator<String> iterator = pendingStops.iterator();
+      while (iterator.hasNext()) {
```

Comment:
seems like these two while loops can be combined -
```java
Iterator<String> iterator = pendingStops.iterator();
while (iterator.hasNext() &&  stopwatch.elapsedTime(TimeUnit.SECONDS) < 3L) {
...
}
```
right?

Reply:
No. The iteration has to restarted from the beginning.

Reply:
oh yeah right